tensorflow/lite/kernels/activations.cc

```diff
@@ -426,13 +425,19 @@ TfLiteStatus TanhPrepare(TfLiteContext* context, TfLiteNode* node) {
         (data->input_left_shift == 0 || data->input_left_shift == 1);
 
     if (!param_scale_pot) {
-      // In case of general scale parameter, we need to do a rescaling.
-      // Magic constant 4096:
-      // We need to scale down to (-2^3, 2^3) / 3 is kInputIntegerBits/ interval
-      // from 16-bit (-2^15, 2^15),
-      // so we need to multiply by
-      // 2^(15 - kInputIntegerBits) = 2^12 = 4096.
-      data->input_multiplier = static_cast<int32_t>(input->params.scale * 4096);
+      // Calculate multiplier to change input scale to 1/(3*4096)
```

Comment:
can you explain more about how the math works here in the comments?

thanks

Reply:
Hi @renjie-liu Sorry for the delay with the reply. I missed the comment.

The initial implementation has been updated to have a general input scale, so we introduced this scaling factor 1/4096 (from 16-bit to [-8, 8]), but the change allows only integer multipliers of the 1/4096 scaling factor. Here we do:
data->input_multiplier = static_cast<int32_t>(input->params.scale * 4096);
and then
int32_t input_data_mul = (input_multiplier > 0) ? input_multiplier : 1;
So, the scaling should be multiple of 1/4096.

This fix is to handle general case.
The number 3.0 in the multiplier comes from here, because the interval is [-10.7, 10.7] instead of [-8, 8]
https://github.com/tensorflow/tensorflow/pull/42671/files#diff-7ae159b53f418105dff8194481058709R66
The numbers are activations_test.cc are identical to values from the calculator with this change.

Reply:
thanks, it would be great to add to the comments